sdk/core/ts-http-runtime/src/auth/credentials.ts

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import type { OAuth2Flow } from "./authFlows.js";
+
+/**
+ * Options used when creating and sending get OAuth 2 requests for this operation.
+ */
+export interface GetOAuth2TokenOptions {
+  /** Abort signal for the request */
+  abortSignal?: AbortSignal;
+}
+
+/**
+ * Options used when creating and sending get bearer token requests for this operation.
+ */
+export interface GetBearerTokenOptions {
+  /** Abort signal for the request */
+  abortSignal?: AbortSignal;
+}
+
+/**
+ * Credential for OAuth2 authentication flows.
+ */
+export interface OAuth2TokenCredential<TFlows extends OAuth2Flow> {
+  /**
+   * Gets an OAuth2 token for the specified flows.
+   * @param flows - The OAuth2 flows to use.
+   * @param options - Options for the request.
+   * @returns - a valid access token which was obtained through one of the flows specified in `flows`.
+   */
+  getOAuth2Token(flows: TFlows[], options?: GetOAuth2TokenOptions): Promise<string>;
+}
+
+/**
+ * Credential for Bearer token authentication.
+ */
+export interface BearerTokenCredential {
+  /**
+   * Gets a Bearer token for the specified flows.
+   * @param options - Options for the request.
+   * @returns - a valid access token.
+   */
+  getBearerToken(options?: GetBearerTokenOptions): Promise<string>;
+}
+
+/**
+ * Credential for HTTP Basic authentication.
+ * Provides username and password for basic authentication headers.
+ */
+export interface BasicCredential {
+  /** The username for basic authentication. */
+  username: string;
+  /** The password for basic authentication. */
+  password: string;
+}
+
+/**
+ * Credential for API Key authentication.
+ * Provides an API key that will be used in the request headers.
+ */
+export interface ApiKeyCredential {
+  /** The API key for authentication. */
+  key: string;
+}
+
+/**
+ * Union type of all supported authentication credentials.
+ */
+export type AuthCredential =
+  | OAuth2TokenCredential<OAuth2Flow>
+  | BearerTokenCredential
+  | BasicCredential
+  | ApiKeyCredential;
+
+/**
+ * Type guard to check if a credential is an OAuth2 token credential.
+ */
+export function isOAuth2TokenCredential(
+  credential: AuthCredential,
+): credential is OAuth2TokenCredential<OAuth2Flow> {
+  return "getOAuth2Token" in credential && credential.getOAuth2Token.length > 0;
```

Comment:
Ican we use credential.type === "oauth2" to discriminate here instead of having to look at the properties?

Reply:
The credential interfaces do not have `type` property

Reply:
We probably don't need the check on the length of the function anymore now that the names are different